File: apps/desktop/src/renderer/src/turns.ts
// One Drone's turns, as rows a person reads: the Called/Answered join, and the
// wire's words where no vocabulary supplies one.
//
// **The join is Bridge's, and that is a decision rather than an oversight.**
// Fleet puts both events on the wire with their call ids because joining them
// there would mean holding a call open until its result arrived — unbounded
// buffering in the loop that advances the Job, in the path that must never
// block. Watching a Job cannot be allowed to change its outcome, so the cost of
// the join lands here, where the worst case is a row that reads "no answer yet".
//
// # What a run cost is drawn here and nowhere else
//
// `ended` carries a Drone run's turn count, its spend and its refusals, and the
// transcript is the only per-run surface Bridge has: a Job that retried has one
// `ended` per Drone, so a figure on the Job — a board column, an outcome row, a
// header fact — would be a total the wire deliberately declines to compute.
// It is also the only surface where the value can be present at all, because it
// arrives on the Observe socket, and the outcome regions of a finished or
// stopped Job are drawn with that socket closed. A labelled blank there would
// be worse than the absence it replaced.
//
// So the run says what it spent on its own last row, and that is still the only
// **per-run** figure there is.
//
// **The Job-level total is served now** — `JobDetail.spend`, since protocol
// 5.5, which `work.ts` draws against the Job's cap. What that closes is a
// person opening one Job and seeing what the whole of it has cost. What is
// still open is comparing Jobs **without** opening one: `JobSummary` carries no
// spend, so the board has no column and cannot have one until it does.

import type { DroneTurn, TurnStep } from "@armada/components";

import type { Turn } from "../../shared/bridge";
import type { JobDetail as JobWhole } from "../../shared/protocol";
import { clock } from "./duration";

/**
 * The rows, with each call carrying whatever came back for it.
 *
 * An `answered` never renders as a row of its own: it is folded into the
 * `called` that shares its id, and two rows would separate a command from its
 * output by everything that happened while it ran. An answer whose call was
 * never seen — the backfill was bounded and cut between them — keeps its own
 * row rather than being dropped, because a dropped row is a gap nobody can see.
 *
 * `whole` is the Job's frozen workflow, and it is what turns a `step_id` into
 * the name the rail already draws. `null` where the detail has not been read;
 * the id renders then, in mono, which is the rail's own fallback.
 */
export function turnsOf(rows: readonly Turn[], whole: JobWhole | null): DroneTurn[] {
  const answers = new Map<string, Answer>();
  for (const row of rows) {
    if (row.saw.event === "answered") answers.set(row.saw.call, { failed: row.saw.failed });
  }

  const seen = new Set<string>();
  const turns: DroneTurn[] = [];
  for (const row of rows) {
    const saw = row.saw;
    if (saw.event === "answered") {
      // Folded into its call, unless nothing here holds that call.
      if (seen.has(saw.call)) continue;
      turns.push({
        id: String(row.seq),
        at: clock(row.ts),
        kind: saw.event,
        step: stepOf(row, whole),
        subject: saw.call,
        answer: ANSWER[saw.failed ? "failed" : "ok"],
      });
      continue;
    }
    if (saw.event === "called") seen.add(saw.call);
    turns.push({
      id: String(row.seq),
      at: clock(row.ts),
      kind: saw.event,
      step: stepOf(row, whole),
      ...bodyOf(saw, answers),
    });
  }
  return turns;
}

/**
 * Which step a row ran under, named the way the rail names one.
 *
 * **A row carrying no step gets none, and never the first one.** Fleet began
 * recording the step on 2026-08-28 and wrote no migration, because the step an
 * older row ran under is unrecoverable — so an absence stays an absence and the
 * pane says so, rather than a transcript claiming a four-step Job happened
 * during step one.
 *
 * **The label is Fleet's and the id is the fallback**, which is `rail.ts`'s
 * rule for the same substitution: `StepDetail.label` is never blank because
 * Fleet substitutes the `step_id` where the workflow declares no name, and a
 * label that *is* the id renders in mono so a reader can tell. A step the
 * frozen workflow does not carry, or a detail nobody has read yet, takes the
 * same fallback — the id is what arrived, and nothing composes a name from it.
 */
function stepOf(row: Turn, whole: JobWhole | null): TurnStep | undefined {
  const id = row.step;
  if (id === undefined) return undefined;
  const label = whole?.steps.find((step) => step.step_id === id)?.label ?? id;
  return { id, label, labelIsAnIdentifier: label === id || undefined };
}

type Answer = { failed: boolean };

type Saw = Turn["saw"];

/**
 * What each kind puts in the row's body.
 *
 * **The kind itself is never translated.** `Saw` is the wire's enum and has no
 * `enum-verbs.toml` rows, so there is no sanctioned verb, glyph or hue for a
 * turn kind and none is written here — the spelling renders, which is
 * recoverable. Reported.
 */
function bodyOf(saw: Saw, answers: Map<string, Answer>): Omit<DroneTurn, "id" | "at" | "kind"> {
  switch (saw.event) {
    case "started":
      return { subject: `${saw.session} · ${saw.model} · ${count(saw.mcp_servers)}` };
    case "called": {
      const answer = answers.get(saw.call);
      const said = answer === undefined ? ANSWER.pending : ANSWER[answer.failed ? "failed" : "ok"];
      // With what the call did in hand, the call id stops being the only thing
      // telling one `Bash` row from the next and stops leading the row. Empty
      // is the wire saying the vocabulary had no name for this tool's
      // arguments — not a field that failed to arrive — and there the id is all
      // there is, so it stays.
      if (saw.detail === "") return { subject: `${saw.tool} · ${saw.call}`, answer: said };
      return { subject: saw.tool, detail: saw.detail, truncated: saw.truncated, answer: said };
    }
    case "said":
      return { said: saw.text };
    case "refused":
      return { subject: `${saw.tool} · ${saw.call}`, said: saw.because };
    // The Drone thinking. Collapsed by the pane into one line carrying the
    // count, because naming these by the decoder's failure to place them
    // describes the plumbing rather than what is happening.
    case "unrecognised":
      return { subject: saw.kind, quiet: true };
    case "unreadable":
      return { subject: saw.line, said: saw.why };
    // What the run cost, on the run's own last row. See `spent` for why the
    // figure hedges and the two counts do not.
    case "ended": {
      const ran = [turned(saw.turns), spent(saw.cost_micros), refused(saw.refusals)];
      return { subject: ran.join(" · ") };
    }
    // An `answered` reaching here is folded above; the arm exists so a kind
    // added to the wire fails to compile rather than rendering as a blank row.
    case "answered":
      return { subject: saw.call };
  }
}

/**
 * What an answer says. Three sentences, because the three are different facts:
 * the tool answered, the tool itself failed — which is not the Drone failing —
 * and the call is still running.
 */
const ANSWER = {
  ok: "Answered.",
  failed: "Answered, and the tool itself failed.",
  pending: "No answer yet.",
} as const;

/** `2 mcp servers`, and `1 mcp server`. */
function count(servers: number): string {
  return servers === 1 ? "1 mcp server" : `${servers} mcp servers`;
}

/** `41 turns`, and `1 turn`. */
function turned(turns: number): string {
  return turns === 1 ? "1 turn" : `${turns} turns`;
}

/**
 * What the run cost, from millionths of a dollar.
 *
 * **Hedged, because a spend figure is not an exit code.** The design contract's
 * P4 spells an estimated value `~$2.40` and never `$2.40`, and every spend
 * rendering in the component library already does: the harness reports the
 * number, nothing here has validated it, and a cost drawn with the authority of
 * a measured value teaches a reader to act on one that may be wrong.
 *
 * **Four places under a cent.** Two would round a run that cost a tenth of a
 * penny to `~$0.00`, which reads as free — and free is the one thing a spend
 * figure must never claim by accident.
 */
function spent(micros: number): string {
  const dollars = micros / 1_000_000;
  return `~$${dollars.toFixed(dollars > 0 && dollars < CENT ? 4 : 2)}`;
}

/** A cent, in dollars. Below this two decimal places round to nothing. */
const CENT = 0.01;

/**
 * How many of the run's calls the harness refused.
 *
 * **Zero is stated rather than left out.** It is the difference between a Drone
 * that was fenced off from what it needed and one that simply went quiet, and a
 * row that says nothing leaves a reader unable to tell "none" from "not
 * reported".
 */
function refused(refusals: number): string {
  if (refusals === 0) return "no calls refused";
  return refusals === 1 ? "1 call refused" : `${refusals} calls refused`;
}

File: crates/ipc/src/tests/turns.rs
//! Which transcript rows a viewer is shown, and what the last one carries.
//!
//! [`Shown`] is the only way onto the socket, so what it refuses is the whole
//! of the withholding — a case here is the design, not a rendering choice.

use crate::{decode, encode, Instant, Saw, Shown, TranscriptRow};

fn row(saw: Saw) -> TranscriptRow {
    TranscriptRow {
        ts: Instant::carried("2026-08-27T14:12:00.000Z"),
        step: None,
        by: crate::Voice::Drone,
        saw,
    }
}

fn ended() -> TranscriptRow {
    row(Saw::Ended {
        turns: 41,
        cost_micros: 1_530_000,
        refusals: 0,
    })
}

/// The Job on 2026-08-27: forty-one turns, a dollar fifty-three, nothing to
/// show for it. Withheld, on the ground that the Job's rail states both — and
/// no rail does, so the numbers were reachable from nowhere.
#[test]
fn what_a_drone_spent_reaches_the_viewer_that_asked_for_its_turns() {
    let shown = Shown::of(ended()).expect("the row a Drone ends on is shown");
    let json = encode(&shown).expect("it encodes");
    assert!(
        json.contains(r#""cost_micros":1530000"#),
        "the spend crosses: {json}"
    );
    assert!(json.contains(r#""turns":41"#), "the turn count crosses");
    let back: Shown = decode("a shown row", json.as_bytes()).expect("it decodes");
    assert_eq!(back.row(), &ended());
}

/// **The narrowing is by field as well as by kind.** A `called` row is shown
/// and the whole of its argument is not, so what a viewer holds is bounded
/// however large the argument was — and the size travels with it, because a row
/// that cannot say how much there is is the dead end this replaced.
#[test]
fn a_shown_call_carries_its_size_and_never_the_argument() {
    let called = row(Saw::Called {
        tool: "Bash".to_string(),
        call: "toolu_01Haa".to_string(),
        detail: "cat <<EOF > notes.md one two three".to_string(),
        truncated: true,
        detail_length: Some(14_320),
        whole: Some("cat <<EOF > notes.md\none\ntwo\nthree\nEOF".to_string()),
    });
    let json = encode(&called).expect("the file holds the whole of it");
    assert!(
        json.contains(r#""whole":"#),
        "the file's own line carries it"
    );

    let shown = Shown::of(called).expect("a call is a row a viewer is shown");
    let json = encode(&shown).expect("it encodes");
    assert!(
        !json.contains(r#""whole":"#),
        "the socket never carries the argument: {json}"
    );
    assert!(
        json.contains(r#""detail_length":14320"#),
        "and it says how much there is, which is what makes the row openable"
    );
}

/// A row from before the file kept a size says so rather than reading as an
/// argument of no length. The same absence `step` carries, for the same reason.
#[test]
fn a_row_written_before_the_size_existed_still_decodes() {
    let line = r#"{"ts":"2026-08-27T14:12:00.000Z","by":"drone","event":"called",
        "tool":"Bash","call":"toolu_1","detail":"ls","truncated":false}"#;
    let back: TranscriptRow = decode("a transcript row", line.as_bytes()).expect("it decodes");
    let Saw::Called {
        detail_length,
        whole,
        ..
    } = back.saw
    else {
        panic!("a call");
    };
    assert_eq!(detail_length, None, "unrecoverable, and not nought");
    assert_eq!(whole, None);
}

/// Showing one kind did not widen the narrowing. `QuotaMoved` is the other
/// side of this and is asserted where it has been since it was written —
/// [`mcp::a_withheld_row_has_no_constructor_and_no_decoder`](super::mcp).
#[test]
fn the_sinks_own_losses_are_still_withheld() {
    let missed = row(Saw::Missed { rows: 12 });
    let json = encode(&missed).expect("the file still holds it");
    assert!(
        Shown::of(missed).is_none(),
        "a viewer's losses are its subscription's, and arrive as its own message"
    );
    assert!(
        decode::<Shown>("a shown row", json.as_bytes()).is_err(),
        "and a peer cannot mint one from the other end either"
    );
}

/// **The three rows the activity log is drawn with, and the voice on each.**
/// The drawing has `Armada`, `Drone` and `Fleet`; the socket carried the middle
/// one, so a surface either invented the other two or explained that they were
/// not on the wire.
#[test]
fn what_armada_said_and_what_fleet_did_reach_the_viewer_with_their_voice() {
    let told = TranscriptRow {
        by: crate::Voice::Armada,
        ..row(Saw::Instructed {
            occasion: "opening".to_string(),
            text: "Implement the route. Done means: a request reaches it.".to_string(),
        })
    };
    let json =
        encode(&Shown::of(told.clone()).expect("Armada's turn is shown")).expect("it encodes");
    assert!(json.contains(r#""by":"armada""#), "whose row it is: {json}");
    assert!(
        json.contains("Done means"),
        "and what the Drone was told, whole and uncut: {json}"
    );
    let back: Shown = decode("a shown row", json.as_bytes()).expect("it decodes");
    assert_eq!(back.row(), &told);

    let checked = TranscriptRow {
        by: crate::Voice::Fleet,
        ..row(Saw::Checked {
            run: crate::CheckRun {
                name: "suite".to_string(),
                outcome: crate::CheckOutcome::from_wire("passed").expect("a registry outcome"),
                expected: None,
                produced: None,
                output_path: None,
            },
        })
    };
    let json = encode(&Shown::of(checked).expect("a Check Fleet ran is shown")).expect("encodes");
    assert!(json.contains(r#""by":"fleet""#), "{json}");
    assert!(json.contains(r#""name":"suite""#), "{json}");

    let produced = TranscriptRow {
        by: crate::Voice::Fleet,
        ..row(Saw::Produced {
            files: vec![crate::ChangedFile {
                path: "crates/api/src/routes.rs".to_string(),
                change: crate::ChangeKind::Modified,
                outside_plan: false,
            }],
        })
    };
    let json =
        encode(&Shown::of(produced).expect("what a step produced is shown")).expect("encodes");
    assert!(json.contains("crates/api/src/routes.rs"), "{json}");
}

/// A row with no voice on it is a Drone's. **Every row written before the field
/// existed is one**, which is what makes the default the truth rather than a
/// convenience.
#[test]
fn a_row_written_before_the_voice_existed_is_the_drones() {
    let old = r#"{"ts":"2026-08-27T14:12:00.000Z","event":"said","text":"reading the file"}"#;
    let back: TranscriptRow = decode("a row", old.as_bytes()).expect("it decodes");
    assert_eq!(back.by, crate::Voice::Drone);
}
